docs/review.md

```diff
@@ -0,0 +1,579 @@
+module Helpers
+
+  def sign_in(user)
+    visit('/sessions/new')
+    fill_in :email, with: user.email
+    fill_in :password, with: user.password
+    click_button "Log in"
+  end
+```
+
+See https://github.com/makersacademy/course/blob/master/pills/spec_helper_methods.md
+
+
+## Ensure asset routes are set correctly
+
+Do we like http://recipes.sinatrarb.com/p/asset_management/sinatra_assetpack ?
+
+Or should we just be telling them to set the public folder correctly
+
+```ruby
+set :public_folder, Proc.new { File.join(root, 'static') }
+```
+
+and for Heroku:
+
+* http://abernardes.github.io/2014/10/24/static-assets-with-rack.html
+* http://stackoverflow.com/a/5509152/316729
+
+add the following to config.ru
+
+```ruby
+use Rack::Static, :urls => ['/stylesheets', '/javascripts'], :root => 'public'
+```
+
+# Step 4: Application code and \*.rb files
+
+## Controllers
+
+### Prefer Sinatra::Application vs Sinatra::Base
+
+TODO work out if we care
+
+* http://www.sinatrarb.com/intro.html#Modular%20vs.%20Classic%20Style
+
+### Be clear about how to use flash.now
+
+Have to register Sinatra Flash before using. Have to use
+
+```
+flash.next[:notice]
+```
+
+and not
+
+```
+flash.now[:notice]
+```
+
+in order to see the message after the redirect.
+
+### Refactor long controller methods
+
+This overly long controller method with business logic spread spread out throughout the method is not ideal.
+
+```ruby
+post '/password_reset' do
+  user = User.first(email: params[:email])
+  if user
+    user.reset_password
+    user.save
+    flash[:info] = 'Check your emails'
+    redirect('/')
+  else
+    flash[:info] = 'No such account on record'
+    redirect('/password_reset')
+  end
+end
+```
+
+It could be refactored like so:
+
+```ruby
+post '/password_reset' do
+  if User.reset_password(email: params[:email])
+    flash[:info] = 'Check your emails'
+    redirect('/')
+  else
+    flash[:info] = 'No such account on record'
+    redirect('/password_reset')
+  end
+end
+```
+
+Note the much shorter method and the business logic all pulled into the User model.
+
+### Split routes into a controller file/many controller files +
+
+Rather than
+
+```ruby
+class Chitter < Sinatra::Application
+  get '/' do
+    redirect '/peeps'
+  end
+
+  get '/peeps' do
+    @peeps = Peep.all
+    erb :'peeps/index'
+  end
+
+  get '/peeps/new' do
+    erb :'peeps/new'
+  end
+end
+```
+
+prefer
+
+```ruby
+# app.rb
+require 'controllers/peep_controller'
+
+class Chitter < Sinatra::Application
+  use Routes::PeepController
+end
+
+# controllers/peep_controller.rb
+module Routes
+  class PeepController < Sinatra::Application
+    get '/' do
+      redirect '/peeps'
+    end
+
+    get '/peeps' do
+      @peeps = Peep.all
+      erb :'peeps/index'
+    end
+
+    get '/peeps/new' do
+      erb :'peeps/new'
+    end
+  end
+end
+```
+
+TODO - do we need the `use` keyword here?
+TODO - watch out for unnecessary inheritance - does PeepController need to extend anything?
```

Comment:
I mean we could do:

```ruby
module Routes
  class PeepController < Sinatra::Application
  end
end
```

or 

```ruby
module Routes
  class PeepController < Sinatra::Base
  end
end
```

or 

```ruby
module Routes
  class PeepController
  end
end
```

and do we need it wrapped in a module?

and how do we pull the routes in:

```ruby
class Chitter < Sinatra::Application
  use Routes::PeepController
end
```

or 

```ruby
class Chitter < Sinatra::Application
  use PeepController
end
```

or 

```ruby
class Chitter < Sinatra::Application
end
require 'peep_controller'  # <-- what we did when you and I were pairing ...
```

or mix and match?